### ISSUE TITLE
"Interpolate values" looks weird

### ISSUE THREAD

**Issue by RubenKelevra** [FL550/dwd_weather#110]
## Version of home_assistant
2023.12.0b1

## Version of the custom_component
2.0.13

## Describe the bug
I've enabled the option "Interpolate values" while setting up the weather and activated a plot of the solar radiation. I've noticed that, while the steps every x hour(s) are gone, the interpolation isn't smooth, but tries only to smooth out half of the time it should. Leading to a half smooth, half stepped graph which looks kinda weird.

I think as solar radiation is kinda smooth on a non-cloudy day the interpolation issue is the most obvious on it.

![Screenshot 2023-12-06 212444_](https://github.com/FL550/dwd_weather/assets/614929/b3b84fbc-ff40-4883-8e77-9ebf0c8ab627)

## Debug log

No debug log due to privacy concerns.

**Comment by FL550** [comment]
I'll have to look into this. Thanks for reporting!

**Comment by FL550** [comment]
As this only occurs in specific points in time, I think it might have something to do with an update of the forecast data. 

The algorithm I use is a linear interpolation, so nothing special. I'll have to further investigate.

**Comment by RubenKelevra** [comment]
Hey @FL550,

Thanks for looking into it.

If I look at the graph it looks to me like there's a shift of half of the interval forward of the interpolated values.

So half of the interval gets no interpolation, then the second half of the interval gets the first part of the interpolation and then the new value creates a 100% gradient in the graph.

**Comment by FL550** [comment]
Hm, I can't confirm this with my setup. I only have a step one time a day, at 11 o'clock. Here is a closer look of one day but this is the same for me for other days:
![Screenshot_20240103_124053](https://github.com/FL550/dwd_weather/assets/36160004/5eb0bdd6-4cea-43bb-8055-9b6c9c6e9766)

Can you please have a closer look at your data so we can track this down?

**Comment by FL550** [comment]
@RubenKelevra any update on this?

**Comment by RubenKelevra** [comment]
Hey @FL550, this graph doesn't look like it's the  same one I'm having troubles with.

I'm seeing the issue especially on the solar radiation one.

Can you check this graph on your end? Maybe the issue is limited to this one?

**Comment by RubenKelevra** [comment]
Well, I think it's also visible on the "cloudyness" graph. But it's so much random data, I'm not sure about it.

Solar radiation however should be smooth, but it's stepped with smoothed out parts in between:

![Screenshot_2024-02-02-18-34-16-756-edit_io homeassistant companion android](https://github.com/FL550/dwd_weather/assets/614929/d5ca65a3-d45d-4d63-8aa7-0cfc3d2fa3e4)

**Comment by FL550** [comment]
@RubenKelevra very strange, I don't have this in my sensor, only a slight step at 11:00 which is due to a data update:
![image](https://github.com/FL550/dwd_weather/assets/36160004/d3dfa675-7d59-40ca-b9ff-a9dc4797856a)

Can you please tell me your station id, so I can have a closer look?

**Comment by t785634** [comment]
My temperature (and also the other values) graph looks very strange, too. Interpolation ist activated.

![IMG_3160](https://github.com/FL550/dwd_weather/assets/125138125/c52589ca-a27e-44eb-a6d1-138ef2b10f38)
![IMG_3161](https://github.com/FL550/dwd_weather/assets/125138125/09ad25e4-488f-4500-90b9-95a85d5b199c)

**Comment by FL550** [comment]
@t785634 @RubenKelevra Sorry for the late reply. Do you have the hourly update enabled? If yes, this might be the cause for this issue.

**Comment by RubenKelevra** [comment]
> @t785634 @RubenKelevra Sorry for the late reply. Do you have the hourly update enabled? If yes, this might be the cause for this issue.

I think you're referring to the interpolation of values, right?

**Comment by FL550** [comment]
@RubenKelevra Yes, the interpolation and the hourly update combined.

**Comment by RubenKelevra** [comment]
> @RubenKelevra Yes, the interpolation and the hourly update combined.

Yeah sure. That's what I'm using.

**Comment by t785634** [comment]
> @t785634 @RubenKelevra Sorry for the late reply. Do you have the hourly update enabled? If yes, this might be the cause for this issue.

Yes, me too. That's what I'm using. Is there a way to use both with correct interpolation?

**Comment by FL550** [comment]
Ok, then this is the error source. However, right now I have no idea how to fix this. Let me explain the problem.

DWD provides the data as hourly values. Lets call the value for the current hour `x0` and the value for the next hour `x1`.
For the interpolation, I use a simple linear function where I gradually shift between the values `x0` and `x1` based on the seconds between these timestamps: `x0 + (x1 - x0) * seconds_difference`

Every new hour, `x0` is replaced by `x1` and the next hourly value is loaded from the cached forecast into `x1` (Normal data update occurs every 6 hours).

However, when the data is refreshed every hour, `x0` is updated to a completely new value, which is most likely different from the old "predicted value in one hour". This produces this visible shift in the data. 

So when for example the old predicted values for the temperatures at 12:00 is `12.0` and at 13:00 is `13.0`, the computed value at 12:59 accorrding to the above formula is approx. `12.98`. If at 13:00 new data is loaded from DWD because of the hourly update and the new, corrected value is now `12.9`, the graph drops from `12.99` to `12.9` instead of rising to the old `13.0`.

**Comment by RubenKelevra** [comment]
I'll give the source code a look later. Thanks for your explanation.

**Comment by FL550** [comment]
I don't think that this will resolve the core issue. However, you are always invited to have a look for improvements.

**Comment by RubenKelevra** [comment]
> Ok, then this is the error source. However, right now I have no idea how to fix this. Let me explain the problem.
> 
> DWD provides the data as hourly values. Lets call the value for the current hour `x0` and the value for the next hour `x1`.
> For the interpolation, I use a simple linear function where I gradually shift between the values `x0` and `x1` based on the seconds between these timestamps: `x0 + (x1 - x0) * seconds_difference`
> 
> Every new hour, `x0` is replaced by `x1` and the next hourly value is loaded from the cached forecast into `x1` (Normal data update occurs every 6 hours).
> 
> However, when the data is refreshed every hour, `x0` is updated to a completely new value, which is most likely different from the old "predicted value in one hour". This produces this visible shift in the data. 
> 
> So when for example the old predicted values for the temperatures at 12:00 is `12.0` and at 13:00 is `13.0`, the computed value at 12:59 accorrding to the above formula is approx. `12.98`. If at 13:00 new data is loaded from DWD because of the hourly update and the new, corrected value is now `12.9`, the graph drops from `12.99` to `12.9` instead of rising to the old `13.0`.

Agreed that's not simple to fix.

The simplest solution would be to ignore the next value, if the hour has already started. And use the updated values for the next hour afterwards.

The next simple solution would be to update the data right before the hour mark, in addition to the fix above. This ensures a low delay for rapidly updated data, as only ~60 minutes delay is introduced, not avg 90 minutes. 

However this would add punctual load onto the DWD infrastructure, as all home assistant instances update at roughly the same time, instead of (presumably currently) just randomly 60 minutes after the boot update.

**Comment by FL550** [comment]
> The simplest solution would be to ignore the next value, if the hour has already started. And use the updated values for the next hour afterwards.

This would mean, that even if there is more precise data available, this data is held back.

> The next simplest solution would be to update the data right before the hour mark, in addition to the fix above. This ensures a low delay for rapidly updated data, as only ~60 minutes delay is introduced, not avg 90 minutes. 

Check for new data is done every 10 minutes, so if the shift occurs during the hour, this means there was a new update from DWD. So nothing can be done about the 90 min delay. DWD updates the data roughly ever 60 minutes.

> However this would add punctual load onto the DWD infrastructure, as all home assistant instances update at roughly the same time, instead of (presumably currently) just randomly 60 minutes after the boot update. 

This is already the case, as soon as new data is available, all clients fetch these new data nearly at the same time.

**Comment by t785634** [comment]
Are there any updates?

**Comment by FL550** [comment]
Please see above, there is currently no good solution for this.

**Comment by RubenKelevra** [comment]
> Please see above, there is currently no good solution for this.

Actually I think there is.

The issue is different than we thought:

Say we got the following values:

08:00 is 1
09:00 is 2

An update is run at 08:30 and the updated value for 09:00 is 2.1.

Then the graph should be increasing to 08:30 with 100% slope and then increase slightly as the value for 09:00 is now higher.

Instead the graph is jumping to 2.1 at 0+the time the update is run.


This is causing this weird steps, if my analysis is correctly.

**Comment by FL550** [comment]
@RubenKelevra Yes, this could smoothen the jumps a bit at the cost of precision. This is however something I can write in the notes so every user is aware of this.

The bigger problem I see is with larger corrections of data like in the image of @t785634 . With your idea implemented this would still result in a zigzag line, as the interpolated value is already above (or below, but just stick to the rising part of temperature in the morning) the new value.

So continuing your example, this would be the case if the new updated value for 09:00 is 1.4 at 08:30.

**Comment by FL550** [comment]
I might have found the issue. Can someone please confirm if this issue is gone now?

**Comment by RubenKelevra** [comment]
@FL550 Well, I can confirm the bug for 2.1.29.

Which version is fixed?

![Screenshot_2025-03-07-14-04-52-778-edit_io.homeassistant.companion.android.jpg](https://github.com/user-attachments/assets/685fc9f2-28ec-465c-9f4c-53def05e683f)

This time a different graph, cause sun radiation is broken:

![Screenshot_2025-03-07-14-05-15-545-edit_io.homeassistant.companion.android.jpg](https://github.com/user-attachments/assets/61bf0281-954e-4739-85c2-78a7521cb7d9)

Did DWD remove support for that?

**Comment by FL550** [comment]
> Which version is fixed?

@RubenKelevra I fixed this (hopefully) in 2.2.0. 

> Did DWD remove support for that? 

Not that I am aware of. Can you please give me your Station ID so I can confirm this in the raw data?

**Comment by RubenKelevra** [comment]
> > Did DWD remove support for that?
> 
> Not that I am aware of. Can you please give me your Station ID so I can confirm this in the raw data?

I've added a different Station ID, and there it's working. Can I send you my Station ID privately somewhere?

**Comment by RubenKelevra** [comment]
Ignore the previous question, I've looked at the raw data myself, it's empty. yay. Guess a sensor on the station broke or something like that 🥲

<details>

```  
                <dwd:Forecast dwd:elementName="RRad1">
                    <dwd:value>         -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -</dwd:value>
                </dwd:Forecast>
                <dwd:Forecast dwd:elementName="Rad1h">
                    <dwd:value>         -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -          -</dwd:value>
                </dwd:Forecast>
```  


</details>

**Comment by RubenKelevra** [comment]
> > Which version is fixed?
> 
> [@RubenKelevra](https://github.com/RubenKelevra) I fixed this (hopefully) in 2.2.0.

I can't test that, because this version does not exist (as download file):

![Image](https://github.com/user-attachments/assets/d60ab036-2d51-4710-a7e3-ce625bfbe00c)

**Comment by FL550** [comment]
> Ignore the previous question, I've looked at the raw data myself, it's empty. yay. Guess a sensor on the station broke or something like that 🥲

Sorry to hear that. Maybe there is another station nearby providing these values?

> I can't test that, because this version does not exist (as download file):

I had to revoke this version due to a missing error check. New version should be available soon.

**Comment by FL550** [comment]
@Markus40 Thanks for investigating! This might be the problem, yes. I'm pretty sure, all internal calculations are in UTC to avoid a discrepancy like this, however I'm not sure if I implemented this for the interpolation. I'll check this as soon as possible, but might take a few days.

**Comment by Markus40** [comment]
Good update! I've been using it for a few days now and the vertical sections have now gone completely.
